File: packages/shared/src/providers.ts
/**
 * Swappable provider interfaces + a registry so concrete implementations plug
 * in during the phase that first needs them (local ↔ free API ↔ paid is a
 * config flip — see STACK.md §4.2). Foundation defines the contracts only.
 *
 * Provider caveat baked in: the Anthropic implementation must NOT send
 * `temperature`/`top_p` (they 400 on Opus 4.8/4.7/Fable 5) — use adaptive
 * thinking. Local/Gemini/Groq use temperature: 0.
 */

import type { Config, EmbeddingProviderName, LLMProviderName, RerankProviderName } from "./config.js";

export interface LLMMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface LLMCompleteOptions {
  system?: string;
  /** When set, the provider must constrain output to this JSON Schema. */
  jsonSchema?: Record<string, unknown>;
  maxTokens?: number;
  /** Sampling temperature (local / Gemini / Groq). Omit for Anthropic. */
  temperature?: number;
  /** Nucleus sampling (local / Gemini / Groq). Omit for Anthropic. */
  topP?: number;
  /**
   * Adaptive-thinking control for the Anthropic provider, which must NOT send
   * temperature/top_p (they 400 on Opus 4.8/4.7/Fable 5).
   */
  thinking?: { type: "adaptive" | "disabled" };
}

export interface LLMProvider {
  complete(messages: LLMMessage[], options?: LLMCompleteOptions): Promise<string>;
}

/**
 * Spreads calls across several LLM providers (round-robin) and fails over to the
 * next on error. Independent free tiers (e.g. Gemini + Groq) share the load, so a
 * run only stalls if EVERY provider is rate-limited at once. Each provider keeps
 * its own internal retry; this wrapper adds the cross-provider failover.
 */
export class MultiLLMProvider implements LLMProvider {
  private next = 0;

  constructor(
    private readonly providers: LLMProvider[],
    private readonly names: readonly string[] = [],
  ) {
    if (providers.length === 0) throw new Error("MultiLLMProvider needs at least one provider");
  }

  async complete(messages: LLMMessage[], options?: LLMCompleteOptions): Promise<string> {
    const n = this.providers.length;
    const start = this.next;
    this.next = (this.next + 1) % n; // round-robin the starting provider per call
    let lastError: unknown;
    for (let i = 0; i < n; i++) {
      const provider = this.providers[(start + i) % n]!; // index always in range
      try {
        return await provider.complete(messages, options);
      } catch (err) {
        lastError = err; // failover to the next provider
      }
    }
    const label = this.names.length ? ` (${this.names.join(", ")})` : "";
    throw new Error(
      `All LLM providers failed${label}: ${lastError instanceof Error ? lastError.message : String(lastError)}`,
    );
  }
}

export interface EmbeddingProvider {
  embed(texts: string[]): Promise<number[][]>;
}

export interface RerankResult {
  index: number;
  score: number;
}

export interface RerankProvider {
  rerank(query: string, documents: string[]): Promise<RerankResult[]>;
}

// --- Registries: phases register concrete implementations here ---
type LLMFactory = (config: Config) => LLMProvider;
type EmbeddingFactory = (config: Config) => EmbeddingProvider;
type RerankFactory = (config: Config) => RerankProvider;

const llmRegistry = new Map<LLMProviderName, LLMFactory>();
const embeddingRegistry = new Map<EmbeddingProviderName, EmbeddingFactory>();
const rerankRegistry = new Map<RerankProviderName, RerankFactory>();

export function registerLLMProvider(name: LLMProviderName, factory: LLMFactory): void {
  llmRegistry.set(name, factory);
}
export function registerEmbeddingProvider(name: EmbeddingProviderName, factory: EmbeddingFactory): void {
  embeddingRegistry.set(name, factory);
}
export function registerRerankProvider(name: RerankProviderName, factory: RerankFactory): void {
  rerankRegistry.set(name, factory);
}

function resolve<N, F>(registry: Map<N, F>, name: N, kind: string): F {
  const factory = registry.get(name);
  if (!factory) {
    throw new Error(`No ${kind} provider registered for '${String(name)}'. Register it before use.`);
  }
  return factory;
}

export function createLLMProvider(config: Config): LLMProvider {
  const names = config.providers.llmList;
  // Single provider: resolve directly (unchanged behaviour).
  if (names.length <= 1) {
    return resolve(llmRegistry, config.providers.llm, "LLM")(config);
  }
  // Several providers: build each with its own model (LLM_MODEL_<PROVIDER>,
  // falling back to LLM_MODEL) and wrap in a round-robin + failover provider.
  const providers = names.map((name) => {
    const model = config.models.llmByProvider[name] ?? config.models.llm;
    const subConfig: Config = {
      ...config,
      providers: { ...config.providers, llm: name },
      models: { ...config.models, llm: model },
    };
    return resolve(llmRegistry, name, "LLM")(subConfig);
  });
  return new MultiLLMProvider(providers, names);
}
export function createEmbeddingProvider(config: Config): EmbeddingProvider {
  return resolve(embeddingRegistry, config.providers.embedding, "embedding")(config);
}
export function createRerankProvider(config: Config): RerankProvider {
  return resolve(rerankRegistry, config.providers.rerank, "rerank")(config);
}

File: services/shared/udaan_shared/providers.py
"""Swappable provider interfaces + registry (Python mirror of
packages/shared/src/providers.ts).

Concrete implementations register in the phase that first needs them. The
Anthropic LLM implementation must NOT send ``temperature``/``top_p`` (they 400
on Opus 4.8/4.7/Fable 5) — use adaptive thinking; local/Gemini/Groq use
temperature 0.
"""

from __future__ import annotations

from typing import Callable, Protocol

from .config import Config


class LLMProvider(Protocol):
    def complete(
        self,
        messages: list[dict[str, str]],
        *,
        system: str | None = None,
        json_schema: dict | None = None,
        max_tokens: int | None = None,
    ) -> str: ...


class EmbeddingProvider(Protocol):
    def embed(self, texts: list[str]) -> list[list[float]]: ...


class RerankProvider(Protocol):
    def rerank(self, query: str, documents: list[str]) -> list[tuple[int, float]]: ...


_llm_registry: dict[str, Callable[[Config], LLMProvider]] = {}
_embedding_registry: dict[str, Callable[[Config], EmbeddingProvider]] = {}
_rerank_registry: dict[str, Callable[[Config], RerankProvider]] = {}


def register_llm_provider(name: str, factory: Callable[[Config], LLMProvider]) -> None:
    _llm_registry[name] = factory


def register_embedding_provider(name: str, factory: Callable[[Config], EmbeddingProvider]) -> None:
    _embedding_registry[name] = factory


def register_rerank_provider(name: str, factory: Callable[[Config], RerankProvider]) -> None:
    _rerank_registry[name] = factory


def _resolve(registry: dict, name: str, kind: str):
    factory = registry.get(name)
    if factory is None:
        raise RuntimeError(f"No {kind} provider registered for '{name}'. Register it before use.")
    return factory


def create_llm_provider(config: Config, *, round_robin: bool = True) -> LLMProvider:
    # LLM_PROVIDER may list several providers ("gemini,groq") for round-robin +
    # failover across independent free tiers. A single name resolves directly.
    # round_robin=False makes a multi-provider list strict priority (try the first,
    # fail over only on error) — used for chat (Groq first, Anthropic backstop).
    names = [n.strip() for n in config.llm_provider.split(",") if n.strip()]
    if len(names) <= 1:
        return _resolve(_llm_registry, names[0] if names else config.llm_provider, "LLM")(config)
    from dataclasses import replace

    from .impl import MultiLLMProvider  # lazy: impl imports this module

    providers = [
        _resolve(_llm_registry, name, "LLM")(
            replace(config, llm_provider=name, llm_model=config.llm_models.get(name, config.llm_model))
        )
        for name in names
    ]
    return MultiLLMProvider(providers, names, round_robin=round_robin)


def create_embedding_provider(config: Config) -> EmbeddingProvider:
    return _resolve(_embedding_registry, config.embedding_provider, "embedding")(config)


def create_rerank_provider(config: Config) -> RerankProvider:
    return _resolve(_rerank_registry, config.rerank_provider, "rerank")(config)
